kernel/thread.c

```diff
@@ -36,6 +36,29 @@ extern struct _static_thread_data _static_thread_data_list_end[];
 	     thread_data < _static_thread_data_list_end; \
 	     thread_data++)
 
+#if defined(CONFIG_THREAD_MONITOR)
+void k_thread_foreach(k_thread_user_cb_t user_cb, void *user_data)
+{
+	struct k_thread *thread;
+	unsigned int key;
+
+	__ASSERT(user_cb, "user_cb can not be NULL");
+
+	key = irq_lock();
+	thread = _kernel.threads;
+	irq_unlock(key);
+
+	while (thread != NULL) {
+		user_cb(thread, user_data);
+
+		key = irq_lock();
+		thread = thread->next_thread;
+		irq_unlock(key);
+	}
+}
+#else
+void k_thread_foreach(k_thread_user_cb_t user_cb, void *user_data) { }
```

Comment:
The k_thread_foreach() might be used for something else than thread state inspection. However it silently fails when CONFIG_THREAD_MONITOR is not set. IMHO we should either make this function always available or add assertion informing user that this function is just a stub when CONFIG_THREAD_MONITOR  is not set.